### ISSUE TITLE
How do I split my data set into a training set, a test set and a validation set

### ISSUE THREAD

**Issue by YINTONG0090** [stellargraph/stellargraph#1959]
### Description

I'm new to graph deep learning.
question1:
Now I want to build a model for link prediction, but I can't understand how the data set was split in the link prediction demo, Can anyone tell me why the training set is split from the test set and not from the original data set? Please remind me if I understand something wrong. Thank you!

question2:
How can I add other indicators, such as recall rate, to the evaluation indicators of the link prediction model?

**Comment by emanoelvianna** [comment]
Hello, this form of division is also causing me some doubts. Shouldn't we use a validation set? In all the examples I ended up finding only the training and testing set.

Perform a division like this wrong:

```
edge_splitter_test = EdgeSplitter(graph)
G_test, edge_ids_test, edge_labels_test = edge_splitter_test.train_test_split(
    p=0.1, method="global", keep_connected=True
)

edge_splitter_train = EdgeSplitter(G_test, graph)
G_train, edge_ids_train, edge_labels_train = edge_splitter_train.train_test_split(
    p=0.1, method="global", keep_connected=True
)


edge_splitter_validation = EdgeSplitter(G_train, graph)
G_validation, edge_ids_validation, edge_labels_validation = edge_splitter_validation.train_test_split(
    p=0.1, method="global", keep_connected=True
)
```

or is it not necessary considering the model in graphs? If so, would you be able to get any reference?

@YINTONG0090 With regard to metrics, Stella relies on tensorflow and keras so you can define metrics in the same way, for example:

```
from tensorflow.keras import layers, optimizers, losses, metrics, Model

model = Model(inputs=x_inp, outputs=prediction)
model.compile(
    optimizer=optimizers.Adam(learning_rate=1e-2),
    loss=losses.binary_crossentropy,
    metrics=["acc", metrics.AUC(curve="ROC", name="auc_roc"), metrics.AUC(curve="PR", name="auc_pr")],
)
```

Thanks

**Comment by YINTONG0090** [comment]
> 你好，这种划分形式也让我有些疑惑。我们不应该使用验证集吗？在所有的例子中，我最终只找到了训练和测试集。
> 
> 执行这样的除法错误：
> 
> ```
> edge_splitter_test = EdgeSplitter(graph)
> G_test, edge_ids_test, edge_labels_test = edge_splitter_test.train_test_split(
>     p=0.1, method="global", keep_connected=True
> )
> 
> edge_splitter_train = EdgeSplitter(G_test, graph)
> G_train, edge_ids_train, edge_labels_train = edge_splitter_train.train_test_split(
>     p=0.1, method="global", keep_connected=True
> )
> 
> 
> edge_splitter_validation = EdgeSplitter(G_train, graph)
> G_validation, edge_ids_validation, edge_labels_validation = edge_splitter_validation.train_test_split(
>     p=0.1, method="global", keep_connected=True
> )
> ```
> 
> 还是没有必要考虑图中的模型？如果是这样，你能得到任何参考吗？
> 
> @YINTONG0090关于指标，Stella 依赖于 tensorflow 和 keras，因此您可以以相同的方式定义指标，例如：
> 
> ```
> from tensorflow.keras import layers, optimizers, losses, metrics, Model
> 
> model = Model(inputs=x_inp, outputs=prediction)
> model.compile(
>     optimizer=optimizers.Adam(learning_rate=1e-2),
>     loss=losses.binary_crossentropy,
>     metrics=["acc", metrics.AUC(curve="ROC", name="auc_roc"), metrics.AUC(curve="PR", name="auc_pr")],
> )
> ```
> 
> 谢谢

thank you! I think there's still a lot to explore (╥﹏╥)

**Comment by YINTONG0090** [comment]
> 你好，这种划分形式也让我有些疑惑。我们不应该使用验证集吗？在所有的例子中，我最终只找到了训练和测试集。
> 
> 执行这样的除法错误：
> 
> ```
> edge_splitter_test = EdgeSplitter(graph)
> G_test, edge_ids_test, edge_labels_test = edge_splitter_test.train_test_split(
>     p=0.1, method="global", keep_connected=True
> )
> 
> edge_splitter_train = EdgeSplitter(G_test, graph)
> G_train, edge_ids_train, edge_labels_train = edge_splitter_train.train_test_split(
>     p=0.1, method="global", keep_connected=True
> )
> 
> 
> edge_splitter_validation = EdgeSplitter(G_train, graph)
> G_validation, edge_ids_validation, edge_labels_validation = edge_splitter_validation.train_test_split(
>     p=0.1, method="global", keep_connected=True
> )
> ```
> 
> 还是没有必要考虑图中的模型？如果是这样，你能得到任何参考吗？
> 
> @YINTONG0090关于指标，Stella 依赖于 tensorflow 和 keras，因此您可以以相同的方式定义指标，例如：
> 
> ```
> from tensorflow.keras import layers, optimizers, losses, metrics, Model
> 
> model = Model(inputs=x_inp, outputs=prediction)
> model.compile(
>     optimizer=optimizers.Adam(learning_rate=1e-2),
>     loss=losses.binary_crossentropy,
>     metrics=["acc", metrics.AUC(curve="ROC", name="auc_roc"), metrics.AUC(curve="PR", name="auc_pr")],
> )
> ```
> 
> 谢谢

edge_splitter_test = EdgeSplitter(graph)
G_test, edge_ids_test, edge_labels_test = edge_splitter_test.train_test_split(
    p=0.1, method="global", keep_connected=True
)

edge_splitter_train = EdgeSplitter(G_test)
G_train, edge_ids_train, edge_labels_train = edge_splitter_train.train_test_split(
    p=0.1, method="global", keep_connected=True
)


edge_splitter_validation = EdgeSplitter(G_train)
G_validation, edge_ids_validation, edge_labels_validation = edge_splitter_validation.train_test_split(
    p=0.1, method="global", keep_connected=True
)

I used a similar approach to yours, only without passing in the 'graph' argument later, and it worked.

**Comment by emanoelvianna** [comment]
Sorry my late, this graph in my code is building before with my dataset, check for exemple:

`graph = StellarGraph({"source": source_data, "target": target_data}, edges, edge_type_column = "edge_type")
`

Best regard,
